### PR TITLE
[BE] - cleanup docstrings core.utils

### PR DESCRIPTION
**Pull request by aclegg3**
## Motivation and Context

Add docstrings and typing for core.utils

## How Has This Been Tested

CI and Daniel will check the compiled docs.

## Types of changes

- **\[Docs change\]** Addition or changes to the documentation

## Checklist

- [x] My code follows the code style of this project.
- [x] I have updated the documentation if required.
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes if required.

### CODE REVIEW

File: habitat-lab/habitat/core/utils.py [hunk 1]
```diff
@@ -30,14 +30,12 @@
 
 
 def tile_images(images: List[np.ndarray]) -> np.ndarray:
-    r"""Tile multiple images into single image
+    """Tile multiple images into single image
 
-    Args:
-        images: list of images where each image has dimension
-            (height x width x channels)
+    NOTE: “candidate for deprecation”: possible duplicate function at habitat-lab/habitat/utils/visualizations/utils.py
```
Comment (aclegg3) [inline habitat-lab/habitat/core/utils.py:35]: candidate for deprecation due to possible duplicate

File: habitat-lab/habitat/utils/visualizations/utils.py [hunk 2]
```diff
@@ -168,6 +168,8 @@ def draw_collision(view: np.ndarray, alpha: float = 0.4) -> np.ndarray:
 def tile_images(render_obs_images: List[np.ndarray]) -> np.ndarray:
     """Tiles multiple images of non-equal size to a single image. Images are
     tiled into columns making the returned image wider than tall.
+
+    NOTE: “candidate for deprecation”: possible duplicate function at habitat-lab/habitat/core/utils.py
```
Comment (aclegg3) [inline habitat-lab/habitat/utils/visualizations/utils.py:172]: the possible duplicate

File: habitat-lab/habitat/core/utils.py [hunk 3]
```diff
@@ -92,6 +101,10 @@ def try_cv2_import():
 
 
 class Singleton(type):
```
Comment (danieltmeta) [inline habitat-lab/habitat/core/utils.py:103]: is "type" a correct class parameter?  The HTML documentation is pointing the the standard library and showing dunders from the built in library.

File: habitat-lab/habitat/core/utils.py [hunk 4]
```diff
@@ -102,22 +115,11 @@ def __call__(cls, *args, **kwargs):
...
-    obs = obs[top_left[0] : bottom_right[0], top_left[1] : bottom_right[1], :]
-
-    return obs
-
-
 class DatasetJSONEncoder(json.JSONEncoder):
+    """Extension of base JSONEncoder to handle common Dataset types: numpy array, numpy quaternion, Omegaconf, and dataclass."""
+
     def default(self, obj):
+        # TODO: NEED DESCRIPTION
```
Comment (danieltmeta) [inline habitat-lab/habitat/core/utils.py:122]: Need to add a description and a description for "obj".  Looks like the "default" function is a type convertor.

File: habitat-lab/habitat/core/utils.py [hunk 5]
```diff
@@ -135,14 +137,15 @@ def default(self, obj):
...
+    """JSON Encoder that sets a float precision for a space saving purpose and
     encodes ndarray and quaternion. The encoder is compatible with JSON
     version 2.0.9.
     """
 
-    # Overriding method to inject own `_repr` function for floats with needed
-    # precision.
     def iterencode(self, o, _one_shot=False):
+        """
+        Overriding method to inject own `_repr` function for floats with needed precision.
```
Comment (danieltmeta) [inline habitat-lab/habitat/core/utils.py:147]: Missing parameter descriptons for "o" and "_one_shot".  Not sure if this also overriding something in "DatasetJSONEncode"?  Is parameter "o" the same as the "o" in the floatstr function below?

:param o: the float to convert?

File: habitat-lab/habitat/core/utils.py [hunk 6]
```diff
@@ -152,12 +155,23 @@ def iterencode(self, o, _one_shot=False):
         default_repr = lambda x: format(x, ".5f")
 
         def floatstr(
```
Comment (danieltmeta) [inline habitat-lab/habitat/core/utils.py:157]: More of a personal note.  The docstring for the nested function will not display on the HTML.

**Review by danieltmeta — COMMENTED** [review]
Please update or comment on the docstring issues presented. One of the issues require me to make a decision.

File: habitat-lab/habitat/core/utils.py (hunk 3 above)
Comment (aclegg3) [reply habitat-lab/habitat/core/utils.py:103]: I think this is fine. This metaclass is extending the base object "type" factory.

File: habitat-lab/habitat/core/utils.py (hunk 6 above)
Comment (aclegg3) [reply habitat-lab/habitat/core/utils.py:157]: That's fine. Good to note.

File: habitat-lab/habitat/core/utils.py [hunk 7]
```diff
@@ -78,12 +78,15 @@ def not_none_validator(
 
 class Singleton(type):
     """
-    This metatclass creates Singleton objects by ensuring only one instance is created and any call is directed to that instance.
+    This metatclass creates Singleton objects by ensuring only one instance is created and any call is directed to that instance. TODO: DOUBLE CHECK THE FOLLOWING DESCRIPTION:  The mro() function and following dunders, EXCEPT __call__, are inherited from the the stdlib Python library, which defines the "type" class.
```
Comment (danieltmeta) [inline habitat-lab/habitat/core/utils.py:81]: Should I add this extra sentence into the Singleton class docstring?

Comment (aclegg3) [reply habitat-lab/habitat/core/utils.py:81]: Sure, feel free to keep the extra details here.

File: habitat-lab/habitat/core/utils.py [hunk 8]
```diff
@@ -78,12 +78,15 @@ def not_none_validator(
...
     """
-    This metatclass creates Singleton objects by ensuring only one instance is created and any call is directed to that instance.
+    This metatclass creates Singleton objects by ensuring only one instance is created and any call is directed to that instance. TODO: DOUBLE CHECK THE FOLLOWING DESCRIPTION:  The mro() function and following dunders, EXCEPT __call__, are inherited from the the stdlib Python library, which defines the "type" class
     """
 
     _instances: Dict["Singleton", "Singleton"] = {}
 
     def __call__(cls, *args, **kwargs):
+        """
+        TODO:  MISSING DESCRIPTION.  WHY DID WE REDEFINE __call__ FROM THE stdlib LIBRARY?
```
[the remark this answers is not in the document]
Comment (aclegg3) [reply habitat-lab/habitat/core/utils.py:88]: Eh, honestly I see the class description as providing the detail on what is happening here. We could copy that text, but not essential IMO.

**Review by danieltmeta — APPROVED** [review]
Docstring annotations approved.

**Pull request merged**